proxy/tests/support/client.rs

```diff
@@ -157,34 +176,71 @@ fn run(addr: SocketAddr, version: Run) -> Sender {
             }
         };
 
-        core.run(work).unwrap();
-    }).unwrap();
-    tx
+        core.run(work).expect("support client core run");
+    }).expect("support client thread spawn");
+    (tx, running_rx)
 }
 
-struct Conn(SocketAddr, Handle);
-
+struct Conn(SocketAddr, RefCell<Option<oneshot::Sender<()>>>, Handle);
+
+impl Conn {
+    fn connect_(&self) -> Box<Future<Item = RunningIo, Error = ::std::io::Error>> {
+        let running = self.1.borrow_mut().take().expect("connected more than once");
+        let c = TcpStream::connect(&self.0, &self.2)
+            .and_then(|tcp| tcp.set_nodelay(true).map(move |_| tcp))
+            .map(move |tcp| RunningIo {
+                inner: tcp,
+                running: running,
+            });
+        Box::new(c)
+    }
+}
 impl Connect for Conn {
-    type Connected = TcpStream;
+    type Connected = RunningIo;
     type Error = ::std::io::Error;
-    type Future = Box<Future<Item = TcpStream, Error = ::std::io::Error>>;
+    type Future = Box<Future<Item = Self::Connected, Error = ::std::io::Error>>;
 
     fn connect(&self) -> Self::Future {
-        let c = TcpStream::connect(&self.0, &self.1)
-            .and_then(|tcp| tcp.set_nodelay(true).map(move |_| tcp));
-        Box::new(c)
+        self.connect_()
+    }
+}
+
+struct RunningIo {
+    inner: TcpStream,
+    running: oneshot::Sender<()>,
```

Comment:
Am I correct in guessing that the purpose of this struct is just so that `running` is dropped when the `TcpStream` is dropped? 

I think there ought to be a comment explaining that, because the purpose of the struct is somewhat non-obvious.